### PR TITLE
Updated documentation for SRT bonding

### PR DESCRIPTION
**Pull request by ethouris**

### CODE REVIEW

File: docs/API-functions.md [hunk 1]
```diff
@@ -400,29 +423,34 @@ streamid or peeraddr.
...
+1. The socket used here may be bound from upside (or binding and connection can
+be done in one function, `srt_connect_bind`) so that it uses a predefined
 network interface or local outgoing port. If not, it behaves as if it was
 bound to `INADDR_ANY` (which binds on all interfaces) and port 0 (which
 makes the system assign the port automatically).
+2. When `u` is a group, then this call can be done multiple times, each time
+for another member connection, and a new member SRT socket will be created
+automatically for every call of this function.
+3. If you want to connect a group to multiple links at once and use blocking
+mode, you might rather want to use `srt_connect_group`.
```
Comment (stevomatthews) [inline docs/API-functions.md]: mode, you might want to use `srt_connect_group` instead.

File: docs/API-functions.md [hunk 2]
```diff
@@ -459,6 +502,126 @@ setting the `SRTO_RENDEZVOUS` option to true, and doing `srt_connect`.
 
 **NOTE:** The port value shall be the same in `local_name` and `remote_name`.
 
+
+Socket group management
+-----------------------
+
+Group types are collected in an `SRT_GROUP_TYPE` type and they are currently:
```
Comment (stevomatthews) [inline docs/API-functions.md]: The following group types are collected in an `SRT_GROUP_TYPE`:

File: docs/API-functions.md [hunk 3]
```diff
@@ -459,6 +502,126 @@ setting the `SRTO_RENDEZVOUS` option to true, and doing `srt_connect`.
...
+```
+int srt_group_data(SRTSOCKET socketgroup, SRT_SOCKGROUPDATA* output, size_t* inoutlen);
+```
+
+* `socketgroup` an existing socket group ID
+* `output` points to an output array
+* `inoutlen` points to a variable set to array's size
+
+This function obtains the current member state of the group specified in
+`socketgroup`. The `output` should point to the array of enough size to get
```
Comment (stevomatthews) [inline docs/API-functions.md]: `socketgroup`. The `output` should point to an array large enough to hold

File: docs/API-functions.md [hunk 4]
```diff
@@ -459,6 +502,126 @@ setting the `SRTO_RENDEZVOUS` option to true, and doing `srt_connect`.
...
+all the elements and `inoutlen` to a variable preset to the size of this array.
```
Comment (stevomatthews) [inline docs/API-functions.md]: all the elements, and `inoutlen` to a variable preset to the size of this array.

File: docs/API-functions.md [hunk 5]
```diff
@@ -459,6 +502,126 @@ setting the `SRTO_RENDEZVOUS` option to true, and doing `srt_connect`.
...
+int srt_connect_group(SRTSOCKET group,
+                      const struct sockaddr* source /*nullable*/, int sourcelen,
+                      SRT_SOCKGROUPDATA name [], int arraysize);
+```
+
+This function does almost the same as calling `srt_connect` (or
+`srt_connect_bind`, if `source` is not NULL) in a loop for every item specified
+in `name` array. However if you did this in a blocking mode, the first call
+to `srt_connect` would block until the connection is established, whereas this
+function blocks until any of the specified connections is established.
```
Comment (stevomatthews) [inline docs/API-functions.md]: This function does almost the same as calling `srt_connect` (or
`srt_connect_bind`, if `source` is not NULL) in a loop for every item specified
in `name` array. However, if you did this in blocking mode, the first call
to `srt_connect` would block until the connection is established, whereas this
function blocks until any of the specified connections is established.

File: docs/API-functions.md [hunk 6]
```diff
@@ -459,6 +502,126 @@ setting the `SRTO_RENDEZVOUS` option to true, and doing `srt_connect`.
...
+```
+
+This function does almost the same as calling `srt_connect` (or
+`srt_connect_bind`, if `source` is not NULL) in a loop for every item specified
+in `name` array. However if you did this in a blocking mode, the first call
+to `srt_connect` would block until the connection is established, whereas this
+function blocks until any of the specified connections is established. 
+
+If you set the group nonblocking mode (`SRTO_RCVSYN` option), there's no
+difference. Note, however, that in this function accepts only groups, not
```
Comment (stevomatthews) [inline docs/API-functions.md]: difference. Note, however, that this function accepts only groups, not

File: docs/API-functions.md [hunk 7]
```diff
@@ -459,6 +502,126 @@ setting the `SRTO_RENDEZVOUS` option to true, and doing `srt_connect`.
...
+
+
+### srt_prepare_endpoint
+
+```
+SRT_API SRT_SOCKGROUPDATA srt_prepare_endpoint(const struct sockaddr* adr, int namelen);
+```
+
+This function turns the given address into `SRT_SOCKGROUPDATA` structure needed by
+`srt_connect_group` function.
```
Comment (stevomatthews) [inline docs/API-functions.md]: This function turns the given address into the `SRT_SOCKGROUPDATA` structure needed 
by the `srt_connect_group` function.

File: docs/API-functions.md [hunk 8]
```diff
@@ -530,14 +694,19 @@ int srt_setsockopt(SRTSOCKET u, int level /*ignored*/, SRT_SOCKOPT opt, const vo
...
+Sets a value for a socket option in the socket or group. The first version
+(`srt_setsockopt`) respects the BSD socket API convention, although the "level"
+parameter is ignored. The second version (`srt_setsockflag`) omits the "level"
+parameter completely.
 
 Options correspond to various data types, so you need to know what data type is 
 assigned to a particular option, and to pass a variable of the appropriate data 
 type with the option value to be set.
 
+Please mind that some of the options can only be set on sockets or only on
```
Comment (stevomatthews) [inline docs/API-functions.md]: Please note that some of the options can only be set on sockets or only on

File: docs/API.md [hunk 9]
```diff
@@ -478,7 +478,10 @@ connected. On binding a socket setting this flag is effective only on this
 socket itself. Note though that there are some post-bound options that have
 important meaning when set prior to connecting.
 
-
+Note that usually options can be set either on a group or on a socket, and
+when set on a group, it will be remembered and then derived by every socket
+that will be automatically created and added to the group. This concerns every
+socket option, unless different rule for that case is declared explicitly.
```
Comment (stevomatthews) [inline docs/API.md]: Note that options are usually set either on a group or on a socket. When set 
on a group, the options will persist. They will be derived by every socket
automatically created and added to the group. This applies to every
socket option, unless a different rule is declared explicitly for a specific case .

File: docs/API.md [hunk 10]
```diff
@@ -946,15 +960,44 @@ pre-1.3.0 version is available only as** `SRTO_LATENCY`.
 | --------------------- | ----- | ------- | ------ | ------ | ------- | ------ |
 | `SRTO_RCVSYN`         |       | pre     | `bool` | true   | true    | false  |
 
-- **[GET or SET]** - Synchronous (blocking) receive mode 
+- **[GET or SET]** - When true, sets blocking mode on reading function, when
```
Comment (stevomatthews) [inline docs/API.md]: - **[GET or SET]** - When true, sets blocking mode on a reading function when

File: docs/API.md [hunk 11]
```diff
@@ -946,15 +960,44 @@ pre-1.3.0 version is available only as** `SRTO_LATENCY`.
...
+reading function will in this case report error `SRT_EASYNCRCV` and return
+immediately. Details depend on the tested entity:
+
+- On a connected socket or group this applies to a receiving function
+(`srt_recv` and others) and a situation when there are no data available for
+reading. The readiness state for this operation can be tested by checking the
+`SRT_EPOLL_IN` flag on the aforementioned socket or group.
+
+- On a freshly created socket or group that is about to be connected to a peer
+listener this applies to `srt_connect` call (and derived), which in
```
Comment (stevomatthews) [inline docs/API.md]: listener this applies to any `srt_connect` (and derived) calls, which in

File: docs/API.md [hunk 12]
```diff
@@ -946,15 +960,44 @@ pre-1.3.0 version is available only as** `SRTO_LATENCY`.
...
+"non-blocking mode" always returns immediately. The connected state for that
```
Comment (stevomatthews) [inline docs/API.md]: "non-blocking mode" always return immediately. The connected state for that

File: docs/API.md [hunk 13]
```diff
@@ -946,15 +960,44 @@ pre-1.3.0 version is available only as** `SRTO_LATENCY`.
...
+state and can be found out only by testing `SRT_EPOLL_ERR` flag.
+
+- On a listener socket this applies to `srt_accept` call. The readiness state
+for this operation can be tested by checking the `SRT_EPOLL_IN` flag on
+this listener socket. This flag is also derived from the listener socket
+by the accepted socket or group, although the meaning of this flag is
+effectively different.
+
+- Note that this flag set on a group is for group only, as it concerns the
+exact receiving operation being done on a group.
```
Comment (stevomatthews) [inline docs/API.md]: - Note that this flag is for a group only, as it concerns a
specific receiving operation being done on a group.

File: docs/API.md [hunk 14]
```diff
@@ -946,15 +960,44 @@ pre-1.3.0 version is available only as** `SRTO_LATENCY`.
...
 | --------------------- | ----- | ------- | ----- | ------ | -------- | ------ |
 | `SRTO_RCVTIMEO`       |       | post    | `int` | msecs  | -1       | -1..   |
 
-- **[GET or SET]** - Blocking mode receiving timeout (-1: infinite)
+- **[GET or SET]** - limit the time up to which the receiving operation will
+block (see `SRTO_RCVSYN` for details), so when this time is exceeded, it
+will behave as if in "non-blocking mode". The -1 value means no time limit.
+
+- Note that this flag set on a group is for group only, as it concerns the
+exact receiving operation being done on a group.
```
Comment (stevomatthews) [inline docs/API.md]: - Note that this flag is for a group only, as it concerns a
specific receiving operation being done on a group.

File: docs/API.md [hunk 15]
```diff
@@ -1081,15 +1124,38 @@ must have a value greater than 1000 - `SRTO_PEERLATENCY`.
 | -------------------- | ----- | ------- | ------ | ------ | -------- | ------ |
 | `SRTO_SNDSYN`        |       | post    | `bool` | true   | true     | false  |
 
-- **[GET or SET]** - Synchronous (blocking) send mode 
+- **[GET or SET]** - When true, sets blocking mode on writing function, when
+it's not ready to perform the operation. When false ("non-blocking mode"), the
```
Comment (stevomatthews) [inline docs/API.md]: - **[GET or SET]** - When true, sets blocking mode on a writing function when
it's not ready to perform the operation. When false ("non-blocking mode"), the

File: docs/API.md [hunk 16]
```diff
@@ -1081,15 +1124,38 @@ must have a value greater than 1000 - `SRTO_PEERLATENCY`.
...
+immediately.
+
+- On a connected socket or group this applies to a sending function
+(`srt_send` and others) and a situation when there's no free space in
+the sender buffer, caused by inability to send all the scheduled data over
+the network. Readiness for this operation can be tested by checking the
+`SRT_EPOLL_OUT` flag.
+
+- On a freshly created socket or group it will have no effect until the socket
+turns into a connected state.
```
Comment (stevomatthews) [inline docs/API.md]: enters  a connected state.

File: docs/API.md [hunk 17]
```diff
@@ -1081,15 +1124,38 @@ must have a value greater than 1000 - `SRTO_PEERLATENCY`.
...
+
+- On a listener socket it will be derived by the accepted socket or group,
+but will have no effect on the listener socket itself.
+
+- Note that this flag set on a group is for group only, as it concerns the
+exact sending operation being done on a group.
```
Comment (stevomatthews) [inline docs/API.md]: - Note that this flag is for a group only, as it concerns a
specific sending operation being done on a group.

File: docs/API.md [hunk 18]
```diff
@@ -1081,15 +1124,38 @@ must have a value greater than 1000 - `SRTO_PEERLATENCY`.
...
 | --------------------- | ----- | ------- | ----- | ------ | -------- | ------ |
 | `SRTO_SNDTIMEO`       |       | post    | `int` | msecs  | -1       | -1..   |
 
-- **[GET or SET]** - Blocking mode sending timeout (-1: infinite)
+- **[GET or SET]** - limit the time up to which the sending operation will
+block (see `SRTO_SNDSYN` for details), so when this time is exceeded, it
+will behave as if in "non-blocking mode". The -1 value means no time limit.
+
+- Note that this flag set on a group is for group only, as it concerns the
+exact sending operation being done on a group.
```
Comment (stevomatthews) [inline docs/API.md]: - Note that this flag is for a group only, as it concerns a
specific sending operation being done on a group.

File: docs/API-functions.md [hunk 19]
```diff
@@ -104,18 +106,29 @@ Creating and configuring sockets
...
+socket family (`AF_INET` or `AF_INET6`). This is now not required; this parameter
+is decided at the call of `srt_conenct` or `srt_bind`.
 
-* `af`: family (either `AF_INET` or `AF_INET6`)
-* `type`, `protocol`: ignored
+* In UDT the `type` parameter was used to specify the file or message mode using
+`SOCK_STREAM` or `SOCK_DGRAM` symbols (with the latter being misleading, as the
+message mode has nothing to do with UDP datagrams and it's rather similar to SCTP
+protocol). In SRT these two modes are available by setting `SRTO_TRANSTYPE` to
+`SRTT_FILE`, and the message mode is set by `SRTO_MESSAGEAPI` option.
```
[the remark this answers is not in the document]
Comment (ethouris) [reply docs/API-functions.md]: I changed it a little bit, so please re-check. The situation is: `SRTO_TRANSTYPE` may be `SRTT_LIVE` (default) or `SRTT_FILE` (this is an option that changes various settings and sets some other options to different defaults - this needs no elaborate description here, just mentioning in short). The `SRTO_MESSAGEAPI` option is then set to default false, which means "stream" mode (TCP-like), and when set to true, it's "message" mode (SCTP-like).

File: docs/API-functions.md (hunk 5 above)
Comment (ethouris) [reply docs/API-functions.md]: Not sure, but I can only see removed "a" in "did this in a blocking mode", am I seeing it right?

File: docs/API.md (hunk 9 above)
Comment (ethouris) [reply docs/API.md]: Not sure, but I used "can" here intentionally. There are some options that make sense (or are allowed to be set) only on group or only on a socket, although most of them can be set on both groups and sockets. Please refer.

File: docs/API.md (hunk 13 above)
Comment (ethouris) [reply docs/API.md]: No, that's not what I meant. This flag can be set on a socket or on a group - not only on a group. The difference is that whereas most of the options can be set on the group only to be derived by the member socket (and are therefore meaningless for the group itself), this option when set on a group, is exactly meaningful for the group, and not derived by the member socket.

File: docs/API.md (hunk 14 above)
Comment (ethouris) [reply docs/API.md]: Same as above.

File: docs/API.md (hunk 17 above)
Comment (ethouris) [reply docs/API.md]: Same above.

File: docs/API.md (hunk 18 above)
Comment (ethouris) [reply docs/API.md]: Same above.

File: docs/API.md (hunk 13 above)
Comment (stevomatthews) [reply docs/API.md]: How about this:
"Note that when this flag is set only on a group, it applies to a
specific receiving operation being done on that group (i.e. it is not 
derived from the socket of which the group is a member)."

File: docs/API-functions.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (stevomatthews) [reply docs/API-functions.md]: How about this:
--BEGIN (line 120)--
protocol). In SRT these two modes are available by setting `SRTO_TRANSTYPE`. The default 
is `SRTT_LIVE`. If, however, you set `SRTO_TRANSTYPE` to `SRTT_FILE` for file mode, you 
can then leave the `SRTO_MESSAGEAPI` option as false (default), which corresponds to 
"stream" mode (TCP-like), or set it to true, which corresponds to "message" mode 
(SCTP-like).
--END--

File: docs/API.md (hunk 13 above)
Comment (stevomatthews) [reply docs/API.md]: @ethouris If you are okay with the change above, I will apply it to the other similar instances below.

File: docs/API-functions.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (ethouris) [reply docs/API-functions.md]: Very good.

File: docs/API.md (hunk 13 above)
Comment (ethouris) [reply docs/API.md]: Yes, that's very good.

File: docs/API.md [hunk 20]
```diff
@@ -1154,8 +1158,9 @@ exact sending operation being done on a group.
 block (see `SRTO_SNDSYN` for details), so when this time is exceeded, it
 will behave as if in "non-blocking mode". The -1 value means no time limit.
 
-- Note that this flag set on a group is for group only, as it concerns the
-exact sending operation being done on a group.
+- Note that when this flag is set only on a group, it applies to a
+specific receiving operation being done on that group (i.e. it is not
```
Comment (stevomatthews) [inline docs/API.md]: specific sending operation being done on that group (i.e. it is not

File: docs/API.md [hunk 21]
```diff
@@ -1141,8 +1144,9 @@ enters a connected state.
 - On a listener socket it will be derived by the accepted socket or group,
 but will have no effect on the listener socket itself.
 
-- Note that this flag set on a group is for group only, as it concerns the
-exact sending operation being done on a group.
+- Note that when this flag is set only on a group, it applies to a
+specific receiving operation being done on that group (i.e. it is not
```
Comment (stevomatthews) [inline docs/API.md]: specific sending operation being done on that group (i.e. it is not

File: docs/API-functions.md [hunk 22]
```diff
@@ -114,15 +114,14 @@ Old and deprecated version of `srt_create_socket`. All arguments are ignored.
...
-
- these two modes are available by setting `SRTO_TRANSTYPE` to
-`SRTT_FILE`, and the message mode is set by `SRTO_MESSAGEAPI` option.
+* In UDT the `type` parameter was used to specify the file or message mode
+using `SOCK_STREAM` or `SOCK_DGRAM` symbols (with the latter being misleading,
+as the message mode has nothing to do with UDP datagrams and it's rather
+similar to the SCTP protocol). In SRT these two modes are available by setting
+`SRTO_TRANSTYPE`. The default is `SRTT_LIVE`. If, however, you set
+`SRTO_TRANSTYPE` to `SRTT_FILE` for file mode, you can then leave the
+`SRTO_MESSAGEAPI option as false (default), which corresponds to "stream" mode
```
Comment (stevomatthews) [inline docs/API-functions.md]: `SRTO_MESSAGEAPI` option as false (default), which corresponds to "stream" mode

File: docs/API.md (hunk 20 above)
Comment (ethouris) [reply docs/API.md]: Ah, my bad blind copy pasting.

**Pull request merged**